### ISSUE TITLE
Gateway mode should support Cloud Foundry ingress routing

### ISSUE THREAD

**Issue by rosenhouse** [istio/istio#2812]
We want to replace the existing ingress routing system in Cloud Foundry (known as "gorouter") with Istio.

The `Gateway` model seems appropriate.

Some requirements:
- define a single port for HTTP and a single port for HTTPS listening
- handle all hosts (shouldn't need to provide a complete list of hosts in my config, it comes from the platform)

Some things we'd like in the future:
- run-time configurable SNI

cc: @shalako @rshriram @ZackButcher 

references:
- [background slides](https://docs.google.com/presentation/d/1gtYNr8te2qeGgWrtH4S073JHmpbkKUkDHWvuLWTSenE/edit#slide=id.g2f930e7539_0_184)
- [background google doc](https://docs.google.com/document/d/10uzw4N11NY_vWxWu2l79Ag96fvPT6w-OGzD_buvILM0/edit)
- [initial PR for gateway](https://github.com/istio/api/pull/254)
- [related issues in istio/api](https://github.com/istio/api/issues?utf8=%E2%9C%93&q=+gateway+)

**Comment by ZackButcher** [comment]
Runtime configuration for SNI is something we'll be shipping in the not too distant future to support other ingress use cases (cc @costinm).

Can you expand some on what you mean by the "handle all hosts" bullet? Today the `hosts` field of the Gateway supports wildcards, so a configuration like `hosts: "*"` is valid.

**Comment by ZackButcher** [comment]
Single port for HTTP/HTTPS aligns well with k8s ingress too - that's the first segment of support for Gateways that I'll be landing, so that aligns well. (The assumption here is that they'll be on ports 80/443 respectively, but that shouldn't be too hard to rework if CF needs something different.)

**Comment by rosenhouse** [comment]
ok, cool.  @rshriram was there something else that you inferred was required of Gateway for Cloud Foundry use case, beyond what has been covered in this issue?

It sounds like we're asking for nothing more than a vanilla Gateway.

**Comment by rosenhouse** [comment]
Ok, here's one thing (copied from [here](https://docs.google.com/presentation/d/1gtYNr8te2qeGgWrtH4S073JHmpbkKUkDHWvuLWTSenE/edit?disco=AAAABoxMzPM&ts=5a691dba&usp=comment_email_document))

> However, if we wanted to support SNI, and allow users to provide custom certs for certain hosts, would we then need to make a separate Gateway resource for each server-cert? Or could a single Gateway resource list many pairs of {hosts, tls}

**Comment by rshriram** [comment]
Here is what I am assuming:

1. Apps created by CF users automatically get a hostname that gets exposed at the ingress. We are talking about 1000s of hosts being added/removed. No single user controls the gateway config.

2. For every request that hits the ingress router with a valid hostname, the request has to be automatically load balanced among the service instances associated with that hostname. (This is essentially an empty route rule of the form

```yaml
hosts:
- myapp.pivotal.io
http:
- route:
   - destination:
           name: myapp.pivotal.io
```

Given that we have 1000s of hosts, we are looking at 1000s of such route rules floating in the system, created automatically by some backend system. (this could be as high as 250K routes?)

3. We also want users to be able to explicitly control the routing. like for requests to myapp.pivotal.io, send 50% to myapp-v1.pivotal.io and other 50% to myapp-v2.pivotal.io. Now, these are likely to be added manually by the user (or through some CI/CD system, etc.). Something has to merge these user generated routes with teh auto-generated routes.

4. Users want to define custom certs for certain domains. You can specify what these custom certs are, using the gateway spec, but since the user is not touching the gateway spec, there has to be something else that takes in the user specified cert, updates the wildcard gateway spec with something like 

```
ports:
- 80
   hosts: *
tls:
- hosts: myapp.pivotal.io
   certs:path-to-my-certs
```

Will this model scale to dynamic add/remove of 1000s of hosts/certs?

Given all this, my suggestion was to provide a few knobs in the gateway spec like this:

```yaml
ports:
- port 80
   get_hosts_from_service_registry: true
   # now hosts spec is not allowed
   # Pilot will automatically generate virtual hosts, and map them to default clusters
tls:
   get_certs_from_secret_service: true # CF folks have to implement Envoy secret discovery service
   secret_service_endpoint: CF-secret-service.foo.bar:8888
   # now manual tls specs are not allowed.
```

A config like above allows the hosts and tls certs in the gateway to be controlled directly by xDS servers without requiring the values to be materialized in the gateway spec. Its a lot less brittle than having to serialize all this in the gateway file and load them up in pilot, and transmit to Envoys. (Keep in mind that CF will not have a k8s api server running in their platform)

**Comment by rshriram** [comment]
cc @louiscryan

**Comment by ZackButcher** [comment]
Yes, you'll either need to create a Gateway resource for them, or update an existing one.

We know there are a lot of use cases already where we'll have a ton of duplicate gateways (e.g. in k8s, Ingress resources all bind to 80/443; we have cases with thousands of ingress resources), so we'll be spending some effort ensuring handling lots of Gateways is efficient.

The goal is that a new Gateway binding to the same ports as existing gateways, with new TLS config, results in just a new entry in the listener's SNI config, a new cluster, and a new route to the cluster.

<hr>

@rshriram re your comments:

I don't think we need a `get_hosts_from_service_registry` option: instead we can use a wildcard for the `hosts` and we'll have a TLS entry (with specific hosts) per app.

**Comment by rshriram** [comment]
> I don't think we need a get_hosts_from_service_registry option: instead we can use a wildcard for the hosts and we'll have a TLS entry (with specific hosts) per app.
We need it to create virtual hosts in Envoy, and the plain route (/) in envoy that maps to the backend cluster that has the hosts for that service. 

The TLS stuff acts only in the SSL context and requires SDS.

**Comment by ZackButcher** [comment]
I was expecting that we'd derive vhosts from the domains in the TLS config (Gateway with "*" host + TLS config for specific hosts should result in clusters being created for the specific hosts).

We'll need to test Envoy to see how it handles lots of certs and rapid rate of change of certs. That's all fairly new code in Envoy, so we'll likely hit some issues there initially.

I do worry about a large number of routes: given that we have multiple use cases with large numbers of simple routes (matches on host, or path prefix no regex), it seems like it might be the time to invest in better-than-linear route selection in Envoy (e.g. trie based lookup for routes/hosts)

**Comment by rshriram** [comment]
So in CF case, people could choose a default cert for several domains (provided by CF), or customize it such that they have their own cert.. And they could choose to have one cert for multiple domains (their custom cert) or a separate cert per domain.

Essentially, domains in CF are kind of cheap (*.bluemix.net, *.pivotal.io, etc..). Create as many as you want and associate them to backend applications.

Bottom line: there is no one to one correspondence between TLS cert and the vhost.

**Comment by louiscryan** [comment]
@rshriram I don't think get_hosts_from_service_registry is the right way to model this. If CF wants to bind an app to the gateway it logically belongs in RouteRule.

Having one RouteRule per app seems reasonable given that it:
- Has the same cardinality as the service registry anyway
- Allows for customization of routing for the app without creating an alternate merge semantic with some other out-of-band feature like the one you suggested

We will have to make this scale as Zack mentioned.

**Comment by rshriram** [comment]
On Sun, Jan 28, 2018 at 4:53 PM Louis Ryan <notifications@github.com> wrote:

> @rshriram <https://github.com/rshriram> I don't think
> get_hosts_from_service_registry is the right way to model this. If CF wants
> to bind an app to the gateway it logically belongs in RouteRule.
>
> Having one RouteRule per app seems reasonable given that it:
>
>    - Has the same cardinality as the service registry anyway
>    - Allows for customization of routing for the app without creating an
>    alternate merge semantic with some other out-of-band feature like the one
>    you suggested
>
> We will have to make this scale as Zack mentioned.
>
>
That being the problem. CF is not going to run k8s api server or etcd. 250K
files holding route rules or 250rules per file doesn’t seem right.

Besides, our design is based on the fundamental assumption that internally,
everything is visible by default and nothing externally. The model CF has
is inverted (apps are visible externally by default). If you turn that
assumption into a config option, you get what I am proposing.

The question is whether we want to be opinionated on the visibility model
or whether we let implementations choose.

**Comment by shalako** [comment]
> 250K files holding route rules or 250rules per file doesn’t seem right.

For what it's worth, I wouldn't expect we'd ever statically configuring Pilot with all these routes. I would think Pilot would discover config from the CF Copilot component, keep it all in memory, and update it as events are received from Copilot.

**Comment by rshriram** [comment]
On Mon, Jan 29, 2018 at 8:06 PM Shannon Coen <notifications@github.com>
wrote:

> 250K files holding route rules or 250rules per file doesn’t seem right.
>
> For what it's worth, I wouldn't expect we'd ever statically configuring
> Pilot with all these routes. I would think Pilot would discover config from
> the CF Copilot component, keep it all in memory, and update it as events
> are received from Copilot.
>

That’s a good point. I had totally forgotten about the in memory config
adapter. That should solve the problem right?

We won’t need this feature I am proposing (pull from service registry
thing).

**Comment by rosenhouse** [comment]
I'd like to revive this discussion a bit by challenging the assumption that an Istio Service should be 1:1 with a Cloud Foundry App.

In K8s, there's a many-to-many relationship between K8s Service and Pod.  That is analogous to CF where we have a many-to-many relationship between CF Route and CF App.

Working from there, an Istio Service is 1:1 with K8s Service, so maybe Istio Service should also be 1:1 with a CF Route.

Next, I'd challenge the assumption that a RouteRule should be required for every service that we want to expose via a Gateway.   I thought a RouteRule was supposed to be an exception to normal traffic flow.  In CF, everything with a CF Route is reachable from outside the cluster.

So I'd go back to @rshriram's suggestion of having a boolean flag on Gateway to enable traffic flow to Services without any route rules.  This is identical to the existing behavior in the Sidecar mode, which provides a default route.  (I might suggest a different name for it, like `expose_services_without_route_rules`) . 

One could still have course-grained control over the exposure by using the Gateway's `Hosts` field with wildcard matching.

Thoughts?

**Comment by ZackButcher** [comment]
It was a conscious design decision to require RouteRules for a service to be exposed via a Gateway. The reasoning behind the difference in philosophy between in-mesh sidecar config (permissive, default RouteRule) and Gateways (strict, requires RouteRule) has a couple of dimensions:
- Gateways typically bridge trust domains, while side cars typically do not
  - e.g. when Gateways act as ingress
- In typical reverse proxy configuration, traffic is routed explicitly
  - i.e. configuration defines the only available routes through the proxy; the configuration does _not_ describe exceptions to normal traffic flow, but defines normal traffic flow

For those reasons we decided to require RouteRules for Gateways. However, I think adding a flag for permissive mode is reasonable given the different use case. There's also a chance we'll need such a flag anyway if we attempt to model sidecars in the mesh as Gateways down the line too.

How exactly do we expect this to work, though? Suppose have permissive mode enabled, we see a request for host `foo.bar.apps` (what do the host names look like in CF routes?), how do we know how to route it? Do we look up the host in a service registry, or perform DNS resolution, or what?

**Comment by rosenhouse** [comment]
The key thing here is whether an Istio Service is a CF App, or a CF Route.

If the Istio Service is a CF App, then your points stand.  The association of an externally-reachable hostname to an app would naturally be an Istio Route Rule.

But if the Istio Service is a CF Route, then the mere existence of such a thing is indication that the user who created that route intends for it to be exposed via ingress.

Either model could work.  The latter model is appealing, because the many-to-many relationship of CF App to CF Route looks, at least to me, somewhat like the many-to-many relationship of K8s Pod-spec to K8s Service.

Here's [a rough sketch of how the latter mapping could work](https://github.com/cloudfoundry/copilot/blob/master/docs/model-mappings.md).  I'm not sure this is the right way to do it, but its worth discussing.

---

I'd like to better understand this part: 

> We wanted to decouple L3/4 proxy configuration from L7 routing configuration

On the mailing list, @rshriram said 

> The kubernetes ingress combines both and it has not worked out well.

Could elaborate here?  Is there someplace I could read more about this?

**Comment by rosenhouse** [comment]
Just chatted with @ZackButcher.  Current plan is:

- Each CF Route gets materialized as an Istio Service
- Istio Gateway does not get the extra "permissive mode" flag.  Instead, CF Platform Adapter for Istio will auto-generate Istio RouteRule for each CF Route

cc @zachgersh @aaronshurley @shalako

**Comment by rshriram** [comment]
completed, pending enablement of CF gateway tests
